File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesActivity.java
package kr.co.yogiyo.rookiephotoapp.diary.main;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

import kr.co.yogiyo.rookiephotoapp.Constants;
import kr.co.yogiyo.rookiephotoapp.R;
import kr.co.yogiyo.rookiephotoapp.camera.CameraActivity;
import kr.co.yogiyo.rookiephotoapp.diary.DiaryEditActivity;
import kr.co.yogiyo.rookiephotoapp.settings.SettingsActivity;

public class DiariesActivity extends AppCompatActivity implements View.OnClickListener {

    // TODO: PAGES의 의미 조사 필요
    public final static int PAGES = 5;
    public final static int LOOPS = 1000;
    public final static int FIRST_PAGE = PAGES * LOOPS / 2;
    public final static Date BASE_DATE = new Date();

    private Calendar calendar;

    private ViewPager viewPager;
    private TextView dateNowPageText;

    public static Calendar getCalendar(int position) {
        Calendar calendar = new GregorianCalendar();
        calendar.setTime(DiariesActivity.BASE_DATE);
        calendar.add(Calendar.MONTH, position - DiariesActivity.FIRST_PAGE);
        return calendar;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_diaries);

        initialize();

        initView();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_start_camera:
                Intent cameraIntent = new Intent(DiariesActivity.this, CameraActivity.class);
                cameraIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(cameraIntent);
                break;
            case R.id.btn_add_diary:
                // TODO: 다이어리 글 추가 화면으로 이동
                Intent intent = new Intent(this, DiaryEditActivity.class);
                intent.putExtra(Constants.DIARY_IDX, -1);
                startActivity(intent);
                break;
            case R.id.btn_convert_layout:
                // TODO: 달력 형태로 변경
                break;
            case R.id.btn_start_settings:
                startActivity(new Intent(DiariesActivity.this, SettingsActivity.class));
                break;
            case R.id.image_show_pre_page:
                viewPager.setCurrentItem(viewPager.getCurrentItem() - 1);
                break;
            case R.id.image_show_next_page:
                viewPager.setCurrentItem(viewPager.getCurrentItem() + 1);
                break;
        }
    }

    private void initialize() {
        calendar = new GregorianCalendar();
        calendar.setTime(new Date());
    }

    private void initView() {
        ImageButton startCameraButton = findViewById(R.id.btn_start_camera);
        ImageButton addButton = findViewById(R.id.btn_add_diary);
        ImageButton convertLayoutButton = findViewById(R.id.btn_convert_layout);
        ImageButton settingsButton = findViewById(R.id.btn_start_settings);
        ImageView showPrePageImage = findViewById(R.id.image_show_pre_page);
        ImageView showNextPageImage = findViewById(R.id.image_show_next_page);
        Toolbar toolbar = findViewById(R.id.toolbar);
        viewPager = findViewById(R.id.viewpager);
        dateNowPageText = findViewById(R.id.text_date_now_page);

        startCameraButton.setOnClickListener(this);
        addButton.setOnClickListener(this);
        convertLayoutButton.setOnClickListener(this);
        settingsButton.setOnClickListener(this);

        showPrePageImage.setOnClickListener(this);
        showNextPageImage.setOnClickListener(this);

        setDateNowPageText(calendar);

        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar == null) {
            throw new NullPointerException();
        }

        DiariesFragmentPagerAdapter diariesFragmentPagerAdapter = new DiariesFragmentPagerAdapter(this, this.getSupportFragmentManager());
        viewPager.setAdapter(diariesFragmentPagerAdapter);
        viewPager.setCurrentItem(FIRST_PAGE);
        viewPager.setOffscreenPageLimit(3);
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int i, float v, int i1) {
                // Do nothing
            }

            @Override
            public void onPageSelected(int position) {
                calendar = getCalendar(position);
                setDateNowPageText(calendar);
            }

            @Override
            public void onPageScrollStateChanged(int i) {
                // Do nothing
            }
        });
    }

    public void setDateNowPageText(Calendar calendar) {
        dateNowPageText.setText(String.format(getString(R.string.text_date_now_page),
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH) + 1));
    }
}

File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/diary/main/DiariesActivity.kt
package kr.co.yogiyo.rookiephotoapp.diary.main

import android.app.DatePickerDialog
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import kotlinx.android.synthetic.main.activity_diaries.*
import kr.co.yogiyo.rookiephotoapp.BaseActivity
import kr.co.yogiyo.rookiephotoapp.GlobalApplication
import kr.co.yogiyo.rookiephotoapp.R
import kr.co.yogiyo.rookiephotoapp.camera.CameraActivity
import kr.co.yogiyo.rookiephotoapp.databinding.ActivityDiariesBinding
import kr.co.yogiyo.rookiephotoapp.diary.DiaryEditActivity
import kr.co.yogiyo.rookiephotoapp.settings.SettingsActivity
import java.util.Date

class DiariesActivity : BaseActivity() {

    private val calendarPickerDialog by lazy {
        YearMonthPickerDialog.newInstance(BASE_DATE.time).apply {
            setListener(DatePickerDialog.OnDateSetListener { _, year, month, _ ->
                this@DiariesActivity.viewpager.currentItem = diariesViewModel.getPositionFromYearMonth(year, month)
            })
        }
    }

    private lateinit var diariesViewModel: DiariesViewModel
    private lateinit var activityDiariesBinding: ActivityDiariesBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initialize()

        initView()
    }

    private fun initialize() {
        diariesViewModel = ViewModelProviders.of(this, DiariesViewModelFactory.getInstance(GlobalApplication.globalApplicationContext))
                .get(DiariesViewModel::class.java).apply {
                    showLoadingView = { showLoading(true) }
                    hideLoadingView = { showLoading(false) }
                }

        activityDiariesBinding = DataBindingUtil.setContentView(this, R.layout.activity_diaries)
        activityDiariesBinding.viewModel = diariesViewModel
    }

    private fun initView() {
        setSupportActionBar(toolbar)

        viewpager.run {
            adapter = DiariesFragmentPagerAdapter(this@DiariesActivity, this@DiariesActivity.supportFragmentManager)
            currentItem = FIRST_PAGE
            offscreenPageLimit = 3
        }

        btn_start_camera.setOnClickListener {
            val cameraIntent = Intent(this@DiariesActivity, CameraActivity::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
            }
            startActivity(cameraIntent)
        }

        btn_date_now_page.setOnClickListener {
            calendarPickerDialog.show(supportFragmentManager, "show_calendar_dialog")
        }

        btn_add_diary.setOnClickListener {
            val intent = Intent(this@DiariesActivity, DiaryEditActivity::class.java).apply {
                putExtra("DIARY_IDX", -1)
            }
            startActivity(intent)
        }

        btn_start_settings.setOnClickListener { startActivity(Intent(this@DiariesActivity, SettingsActivity::class.java)) }

        btn_show_pre_page.setOnClickListener { viewpager.currentItem -= 1 }

        btn_show_next_page.setOnClickListener { viewpager.currentItem += 1 }
    }

    companion object {
        const val PAGES = 5
        const val LOOPS = 1000
        const val FIRST_PAGE = PAGES * LOOPS / 2
        var BASE_DATE = Date()
    }
}